### PR TITLE
Update gateway port to service port before gateway merging.

### PR DESCRIPTION
**Pull request by bianpengyuan**
fix https://github.com/istio/istio/issues/29291.

### CODE REVIEW

**Comment by hzxuzhonghu** [comment]
Let's say i have a ingress gateway svc with:

port   targetPort

80     8080
8080 8088

and gateway server's port specified

8080

IIUC,  without this pr, the traffic goes to target port 8088, 
with this patch applied, it goes to target port 8080

**Comment by bianpengyuan** [comment]
@hzxuzhonghu Good catch! I updated the code to skip rewriting if it is ambiguous on which port the server port maps to.

**Comment by hzxuzhonghu** [comment]
I think this direction is right, it means for gateway the svc port takes priority over the target port. cc @howardjohn @ramaraochavali WDYT

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -1707,11 +1715,27 @@ func (ps *PushContext) mergeGateways(proxy *Proxy) *MergedGateway {
...
+					// Check if the gateway server port is also defined as a service port, if so skip rewriting since it is
+					// ambiguous on whether the server port points to service port or target port.
+					if _, ok := servicePorts[s.Port.Number]; ok {
+						continue
+					}
 
+					// The gateway server is defined with target port. Convert it to service port before gateway merging.
+					// Gateway listeners are based on target port, this prevents duplicated listeners be generated when build
+					// listener resources based on merged gateways.
+					s.Port.Number = servicePort
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go:1733]: I think you are mutating it for all references. need a deep copy probably

Comment (howardjohn) [reply pilot/pkg/model/push_context.go:1733]: we do similar with VS when we resolve hosts to FQDN

**Review by howardjohn — COMMENTED** [review]
This also fixes https://github.com/knative/serving/issues/10160, right?

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:1733]: yeah, it is mutating the k8s informer cache, which is not encouraged

**Comment by bianpengyuan** [comment]
> This also fixes knative/serving#10160, right?

Yeah there is a test case for it.

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (bianpengyuan) [reply pilot/pkg/model/push_context.go]: Done.

File: releasenotes/notes/gateway-target-port.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: networking
+issue:
+  - 29291
+
+releaseNotes:
+- |
+  **Fixed** an issue where duplicated listeners are generated when Gateway server ports use mix of service port and target port.
```
Comment (ericvn) [inline releasenotes/notes/gateway-target-port.yaml:9]: Maybe **Fixed** an issue where duplicated listeners are generated when Gateway server ports use **a** mix of service target ports

Comment (bianpengyuan) [reply releasenotes/notes/gateway-target-port.yaml:9]: The issue happens if user have two gateways, one points to service port (say 80), the other one points to the target port of that service port (say 8080). So I think the current way to describe the issue is more accurate?

**Review by howardjohn — APPROVED** [review]

**Comment by bianpengyuan** [comment]
Looks like this PR has already been merged (https://github.com/istio/istio/commit/ae8b0e2cf984f13224b35423edb52f25832ac616). I guess it is because of github glitch. Close this now.

**Pull request closed without merging**